### ISSUE TITLE
Analyzer to remove unnecessary null-forgiving operators

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#34714]
Especially as the compiler improves its nullability analysis and reduces the number of false positive warnings for potential null dereferences (e.g. via improved attribution from https://github.com/dotnet/roslyn/issues/26761), we are going to find ourselves with unnecessary `!`s strewn throughout the codebase.  It would be helpful if there were an analyzer + auto-fix that would spot the unnecessary usage and enable automatically removing it.

cc: @jaredpar, @dotnet/nullablefc

**Comment by jaredpar** [comment]
Related to #25372

**Comment by stephentoub** [comment]
Thanks, @jaredpar.  Should I close this one in favor of #25372?

**Comment by sharwell** [comment]
Duplicate of #25372

**Comment by sharwell** [comment]
I'm reopening this.

The compiler diagnostic (#25372) would always be a hidden diagnostic, and would report diagnostics even in generated code. This issue covers the corresponding analyzer diagnostic, which would have configurable severity, an associated code fix, and _not_ report diagnostics in generated code.

**Comment by YairHalberstadt** [comment]
@jnm2 

Didn't you write a tool to do this?

**Comment by jnm2** [comment]
Yes, more details in https://github.com/dotnet/roslyn/issues/41574

**Comment by jnm2** [comment]
ReSharper now has this feature: https://www.jetbrains.com/resharper/whatsnew/2021-2/#nrt

**Comment by CyrusNajmabadi** [comment]
> What does block adding this feature please?

There is nothing blocking the addition of this feature :-)

**Comment by CyrusNajmabadi** [comment]
It would need a compiler diagnostic. But anyone is free to contribute that.

**Comment by mikernet** [comment]
As some food for thought on approaches, I am finding this extension wonderful for managing null forgiveness in a project: https://github.com/tom-englert/Nullable.Extended

**Comment by CyrusNajmabadi** [comment]
@Neme12 prs welcome.

**Comment by RikkiGibson** [comment]
It looks like deciding whether a `!` is unnecessary, when the `!` changes the type of the expression, is tricky. https://github.com/dotnet/roslyn/issues/25372#issuecomment-639647979

I think it would be worthwhile to start with reporting on cases where `e!` *doesn't* change the type of the expression, for example when `e` itself has non-null flow state, and when its removal doesn't result in a nested conversion warning (like in the `List<string> li = new List<string?>()!;` case).

I would be happy to review a community PR which reports for the simple case.

Then, more complex cases involving changing flow state, weird cases where removing a subset but not single operators avoids warnings, etc., can be addressed later. I feel as though there has to be a way to analyze code, as though it contains no `!`s, and decide which places need `!`s, without editing and re-binding an exponential number of times. But I could be wrong.

**Comment by alexandrehtrb** [comment]
What I understood is that it's feasible to run this analyzer when the target type is known, but not when is expression-determined (aka `var`), right?

Maybe we could have a `var?` keyword to indicate that it's nullable; `var` is non-nullable. Of course, this would be a big change, but it would be helpful to diagnose and prevent nullability issues.

**Comment by CyrusNajmabadi** [comment]
@znakeeye 

> Can't we just have a setting

You can whip up an analyzer for this in a few minutes.

**Comment by mikernet** [comment]
@znakeeye Did you look at what I linked above? https://github.com/tom-englert/Nullable.Extended

It has an analyzer you can use. You can set it to warn for any unjustified null forgiveness operators. I think that should essentially meet your needs.

**Comment by Suchiman** [comment]
> You can set it to warn for any unjustified null forgiveness operators

As far as i understand the description, `unjustified` in this case means that you did not explain with a comment, why you added the `!`, not that it is unnecessary.

**Comment by mikernet** [comment]
The person I am replying to was asking how to make usage of the operator a warning. The suggestion I gave does that.

**Comment by mikernet** [comment]
The VS extension also handles unnecessary null forgiveness if you want that.
